EKF Miscellaneous Improvements and Bug Fixes

priseborough opened the pull request:
Allows use of range beacons simultaneously with GPS and optical flow
Fixes an error in the way that the Pozyx beacon system output is rotated into the local NED earth frame.
Makes the sensor buffer lengths adapt to the time delays to reduce output predictor errors
Uses the averaged, but unfiltered airspeed to reduce the effective time delay.
Fixes a potential bug that could allow the filter to start before the required sensor buffers are filled
Uses the time delay published by the GPS driver, instead of a value specified by a EKF parameter.
Updates EKF3 to allow copters to fly in non position aiding modes without a magnetometer.

TODO:

Add GPS_DELAY and GPS_DELAY2  parameters to the GPS library as discussed in https://github.com/ArduPilot/ardupilot/pull/5405

OXINARF commented on libraries/AP_NavEKF3/AP_NavEKF3_core.cpp:
```diff
@@ -47,34 +46,62 @@ bool NavEKF3_core::setup_core(NavEKF3 *_frontend, uint8_t _imu_index, uint8_t _c
+    // Calculate the expected EKF time step
+    if (_ahrs->get_ins().get_sample_rate() > 0) {
+        dtEkfAvg = 1.0f / _ahrs->get_ins().get_sample_rate();
+        dtEkfAvg = fmaxf(dtEkfAvg,EKF_TARGET_DT);
```
Any reason to not use our MAX definition?

priseborough replied:
None, is there a benefit?

priseborough replied:
I went through the EKF3 code and both fmaxf() and MAX() functions are being used for floating point constraints in different places. I want to be consistent, but before changing them all to one or the other, would like to know what the advantages and disadvantage are.

OXINARF commented on libraries/AP_NavEKF3/AP_NavEKF3.cpp:
```diff
@@ -219,7 +214,7 @@ const AP_Param::GroupInfo NavEKF3::var_info[] = {
     // @Param: HGT_DELAY
     // @DisplayName: Height measurement delay (msec)
-    // @Description: This is the number of msec that the Height measurements lag behind the inertial measurements.
+    // @Description: This is the number of msec that the Height measurements lag behind the inertial measurements. The autpilot should be restarted after adjusting this parameter.
```
Instead of all these comments in the description it should instead be added `@RebootRequired: True`

priseborough replied:
will fix

priseborough commented on libraries/AP_NavEKF3/AP_NavEKF3_RngBcnFusion.cpp:
```diff
@@ -25,11 +25,27 @@ void NavEKF3_core::SelectRngBcnFusion()
+                FuseRngBcn();
+            } else {
+                // If we are using GPS, then GPS is the primary reference, but we continue to use the beacon data
+                // to calculate an indpendant position that is used to update the beacon position offset if we need to
```
wil fix

priseborough commented on libraries/AP_NavEKF3/AP_NavEKF3_core.h:
```diff
@@ -894,6 +894,19 @@ class NavEKF3_core
+    enum resetDataSource {
+                    DEFAULT=0,      // Use best available data
+                    GPS=1,          // Use GPS if available
+                    RNGBCN=2,       // Use beacon range data if available
```
will fix

OXINARF commented on libraries/AP_NavEKF3/AP_NavEKF3_MagFusion.cpp:
```diff
@@ -274,10 +274,11 @@ void NavEKF3_core::SelectMagFusion()
+    // airborne. For other platform types we do this all the time.
     if (!use_compass()) {
-        if (onGround && (imuSampleTime_ms - lastSynthYawTime_ms > 1000)) {
+        if ((onGround || assume_zero_sideslip()) && (imuSampleTime_ms - lastSynthYawTime_ms > 140)) {
```
According to the comment I think this should be:

`onGround || !assume_zero_sideslip()`

Either we are on the ground and do it for all vehicles or we aren't and we do it for non-plane vehicles.

priseborough replied:
will fix

WickedShell commented on libraries/AP_NavEKF3/AP_NavEKF3_core.cpp:
```diff
@@ -47,34 +48,75 @@ bool NavEKF3_core::setup_core(NavEKF3 *_frontend, uint8_t _imu_index, uint8_t _c
+            GCS_MAVLINK::send_statustext_all(MAV_SEVERITY_INFO, "EKF3 waiting for GPS config data");
+        }
         return false;
+    } else if (!_ahrs->get_gps().all_configured()) {
```
I can't see a reason to have this case. We should always succeed at configuring a GPS, unless the user requested to not configure it, in that case we should never opt out of the time delay, and instead should always be waiting until we succeed configuration.

priseborough replied:
Agree - provided the GPS library all_configured() function does not return false indefinitely for a GPS of an unknown type.

priseborough replied:
OK, will remove the timeout and also add the delay configuration parameters to the GPS library.

WickedShell commented:
It doesn't, it will always say it's configured unless it's a ublox and
working on the configuration.

On Dec 28, 2016 4:04 PM, "Paul Riseborough" <notifications@github.com>
wrote:

> *@priseborough* commented on this pull request.
> ------------------------------
>
> In libraries/AP_NavEKF3/AP_NavEKF3_core.cpp
> <https://github.com/ArduPilot/ardupilot/pull/5407>:
>
> >          return false;
> +    } else if (!_ahrs->get_gps().all_configured()) {
>
> Agree - provided the GPS library all_configured() function does not return
> false indefinitely for a GPS of an unknown type.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/ArduPilot/ardupilot/pull/5407>, or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAipiNSd1-cEk6ZDiForKg7PJ_jf5VVuks5rMs7GgaJpZM4LQcWt>
> .
>

priseborough commented:
I flight tested this on a copter  with px4flow sensor enabled and all COMPASS_USE parameters set to 0.

Flight was stable with some heading drift at the start that disappeared as it did manage to learn the small Z gyro offset with some low speed manoeuvring. Here are the gyro bias estimates:
![screen shot 2016-12-29 at 9 45 32 am]
(https://cloud.githubusercontent.com/assets/3596952/21533193/37174a96-cdac-11e6-82a2-f89729036146.png)

Flow innovations looked normal:
![screen shot 2016-12-29 at 9 50 32 am](https://cloud.githubusercontent.com/assets/3596952/21533222/6fb7b4ee-cdac-11e6-9cb4-6578258bc9d2.png)

Magnetometer innovations were zero as expected:
![screen shot 2016-12-29 at 9 52 35 am](https://cloud.githubusercontent.com/assets/3596952/21533237/8c3cd720-cdac-11e6-8d8b-a501003277d3.png)

OXINARF commented on libraries/AP_GPS/AP_GPS.cpp:
```diff
@@ -196,6 +196,23 @@ const AP_Param::GroupInfo AP_GPS::var_info[] = {
+    // @Description: Controls the amount of GPS  measurement delay that the autopilot compensates for. Set to zero to use the default delay for the detected GPS type.
+    // @Units: msec
+    // @Min: 0
+    // @Max: 250
```
Not Min/Max but `@Range: 0 250`

OXINARF commented on libraries/AP_GPS/AP_GPS.cpp:
```diff
@@ -196,6 +196,23 @@ const AP_Param::GroupInfo AP_GPS::var_info[] = {
+    // @Description: Controls the amount of GPS  measurement delay that the autopilot compensates for. Set to zero to use the default delay for the detected GPS type.
+    // @Units: msec
+    // @Min: 0
+    // @Max: 250
```
Same

rmackay9 commented on libraries/AP_GPS/AP_GPS.cpp:
```diff
@@ -196,6 +196,21 @@ const AP_Param::GroupInfo AP_GPS::var_info[] = {
+    // @Param: DELAY_MS
+    // @DisplayName: GPS delay in milliseconds
+    // @Description: Controls the amount of GPS  measurement delay that the autopilot compensates for. Set to zero to use the default delay for the detected GPS type.
+    // @Units: msec
```
@Units: milliseconds would be more consistent with what we call it in other libraries.

priseborough replied:
I did a search through the code found the following usages in different places

milli-seconds
milliseconds
Milliseconds
msec (EKF libraries)

I will add patches to change them all to milliseconds

priseborough commented:
@roger-zhao I believe I have addressed review comments, so there is nothing more I can do from my end to get this in.

This PR is not required to fly in loiter mode with only GPS. Ardupilot has been able to do that for years.

roger-zhao commented:
@priseborough , Sorry, Paul, It's my fault not express clearly what I thought, actually I mean if I can fly copter without compass(mightbe just initialize the heading when starting) and only with a good GPS in Loiter mode according to this PR, :). Again, tks!

priseborough commented:
@OXINARF the perf counter is  in the calling function for https://github.com/ArduPilot/ardupilot/blob/master/libraries/AP_NavEKF3/AP_NavEKF3_MagFusion.cpp#L437 , so no end required. 
Will push a patch to fix the other.

priseborough commented on libraries/AP_GPS/AP_GPS.cpp:
```diff
@@ -858,3 +873,21 @@ void AP_GPS::inject_data_all(const uint8_t *data, uint16_t len)
+    } else if (drivers[instance] == nullptr || state[instance].status == NO_GPS) {
+        // the user has not specified a value and we cannot determine it from the GPS type
+        // so return a default delay of 1 measurement interval
+        return 0.001f * (float)_rate_ms[instance];
```
Yes, this was written before those changes. Will fix.

OXINARF commented:
> @OXINARF the perf counter is in the calling function for https://github.com/ArduPilot/ardupilot/blob/master/libraries/AP_NavEKF3/AP_NavEKF3_MagFusion.cpp#L437 , so no end required.

Sorry, you are right, I probably looked a second time at the EKF2 instead of looking at the EKF3.

When you push the fixes, please make a comment and I'll merge it. Thanks!

priseborough commented on libraries/AP_GPS/AP_GPS.h:
```diff
@@ -310,7 +310,8 @@ class AP_GPS
 
     // the expected lag (in seconds) in the position and velocity readings from the gps
-    float get_lag() const { return 0.2f; }
+    float get_lag(uint8_t instance) const;
```
I will have a look at it, but it would involve making changes in multiple files for no benefit other than a few bytes of storage.

priseborough replied:
I had a go at it, but ran into compiler errors associated with use of the override attribute on the get_lag function that I did not understand, so will tackle that separate to this PR.

priseborough replied:
The branch i attempted this with is here: https://github.com/priseborough/ardupilot/tree/temp

priseborough commented:
I've made changes for the review comments and rebased.

OXINARF commented on libraries/AP_GPS/AP_GPS.cpp:
```diff
@@ -884,10 +884,6 @@ float AP_GPS::get_lag(uint8_t instance) const
-    } else if (drivers[instance] == nullptr || state[instance].status == NO_GPS) {
-        // the user has not specified a value and we cannot determine it from the GPS type
-        // so return a default delay of 1 measurement interval
-        return 0.001f * (float)_rate_ms[instance];
```
@priseborough This is actually wrong, we should check for a non-existent driver in the requested instance (it would be a wrong use of the function from the caller but we should prevent it here). I think we should return a default 200 ms in that case.
Would you prefer to use the _rate_ms in case a backend exists but doesn't have it's own defined lag (so every GPS except Ublox)? I can do that if you want.

priseborough replied:
If the user has specified a delay for that instance, then we should use it, otherwise, given we don't know what type of GPs it is, we should set the delay to one measurement interval (200msec for a 5Hz GPS).

OXINARF commented:
@priseborough Can you check that you agree with https://github.com/ArduPilot/ardupilot/compare/master...OXINARF:review/5407?expand=1? It is basically this PR plus your changes to return lag in msec and my last commit changing the backend to return 1 measurement interval when the driver doesn't have a specific lag.

priseborough commented:
@OXINARF How about we get this PR in with the changes required to make it work and then deal with 'nice to haves' like changing the units of lag from sec to msec in a later PR. I don't have time at the moment to deal with any more changes.

rmackay9 commented:
Merged, thanks for the code and reviews!
I think @OXINARF and @priseborough are going to handle the remaining questions/comments as separate PRs.
